[ANT-3642] 5.6, 5.7 max, min and pow operators

a-zakir opened the pull request.

a-zakir committed:
max

a-zakir committed:
restore grammar

a-zakir committed:
use FunctionNode

a-zakir committed:
fix test

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
max operator

a-zakir committed:
update

a-zakir committed:
update grammar

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
fix max

a-zakir committed:
update grammar

a-zakir committed:
use ^ instead of pow

a-zakir committed:
fix

a-zakir committed:
merge dev

a-zakir committed:
fix merge conflicts

a-zakir committed:
upd

a-zakir committed:
handle min

a-zakir committed:
upd

a-zakir committed:
try fix

a-zakir committed:
add tests

flomnes commented on src/expressions/visitors/LinearityVisitor.cpp:
```diff
@@ -120,18 +123,103 @@ LinearStatus LinearityVisitor::visit([[maybe_unused]] const Nodes::AllTimeSumNod
+    }
+    EvalVisitor evalVisitor(optimEntityContainer_, fillContext_, component_);
+    auto exponentValue = evalVisitor.dispatch(operands[1]).valueAsDouble();
+    if (std::abs(exponentValue) < 1e-12)
```
Magic value here

a-zakir replied:
Yes i don’t know what epsilon should i use

flomnes commented on src/expressions/visitors/LinearityVisitor.cpp:
```diff
@@ -120,18 +123,103 @@ LinearStatus LinearityVisitor::visit([[maybe_unused]] const Nodes::AllTimeSumNod
+        return LinearStatus::CONSTANT;
+    }
+    // If base is linear and exponent is 1 the result is linear
+    if (base == LinearStatus::LINEAR && std::abs(exponentValue - 1) < 1e-12)
```
Magic value here

a-zakir replied:
replace with DEFAULT_THRESHOLD

flomnes commented on src/io/inputs/model-converter/convertorVisitor.cpp:
```diff
@@ -389,32 +397,137 @@ std::any ConvertorVisitor::visitDual(ExprParser::DualContext* context)
+}
+template<class T>
+std::any ConvertorVisitor::processPower(std::vector<T*> exprContexts,
+                                        const std::function<std::string()>& toStringTreeCallBack)
```
Not sure if we really need a callback in this case. Since it accepts no arguments, it seems like a `std::string` would be enough.

Also, a `std::function` can be empty, `operator()` throws `std::bad_function_call` if `*this` does not store a callable function target, i.e. `!*this == true`, see [`operator()`](https://en.cppreference.com/w/cpp/utility/functional/function/operator().html).

a-zakir replied:
Most Time it won’t be used that’s why it is not a string.

a-zakir replied:
don't worry about std::function being empty, the function is used correctly and privately two times.

flomnes commented on src/io/inputs/model-converter/convertorVisitor.cpp:
```diff
@@ -389,32 +397,137 @@ std::any ConvertorVisitor::visitDual(ExprParser::DualContext* context)
+    {
+        throw std::invalid_argument("bad power expression, the exponent is invalid in "
+                                    + toStringTreeCallBack());
+    }
```
Not sure this can happen

a-zakir replied:
Let me Check. However this is helpful in case of syntaxe error, could be used elsewhere

a-zakir replied:
removed

flomnes commented on src/io/inputs/model-converter/convertorVisitor.cpp:
```diff
@@ -389,32 +397,137 @@ std::any ConvertorVisitor::visitDual(ExprParser::DualContext* context)
+    {
+        throw std::invalid_argument("min operator expect at least 2 operands got "
+                                    + std::to_string(nodes.size()));
+    }
```
`min(a, b, c)` ?

a-zakir replied:
Bad copy should be updated

flomnes commented on src/modeler-optimisation-container/include/antares/modeler-optimisation-container/OptimEntityContainer.h:
```diff
@@ -74,6 +74,17 @@ class OptimEntityContainer
+        bool isVarScenarioDep = modelVar.IsScenarioDependent();
+        return Optimisation::convertToTimeIndex(isVarTimeDep, isVarScenarioDep);
+    }
+
```
Unused ?

flomnes replied:
Unused.

a-zakir replied:
removed

flomnes commented on src/solver/optim-model-filler/TimeDependentLinearExpression.cpp:
```diff
@@ -210,6 +210,28 @@ TimeDependentLinearExpression TimeDependentLinearExpression::operator/(
+    TimeDependentLinearExpression out(*this);
+    for (std::size_t t = 0; t < size(); t++)
+    {
+        out[t] = out[t].min(other[t]);
```
It seems like this has useless copies (`operator=`). How about

```cpp
// (out[t] < other[t]) ? NOOP() : out[t] = other[t]
out[t].min(other[t]); // applyMin if you prefer
```
?

flomnes replied:
Also you may need to add a few
```cpp
if (other.size() > out.size())
{
    out.expandTo(other.size());
}
```
in order to resize `*this` if `other` is larger.

a-zakir replied:
useless and and inefficient:
see 
```
const LinearExpression& TimeDependentLinearExpression::operator[](std::size_t idx) const
{
    if (isConstant())
    {
        return v_[0];
    }
    else
    {
        return v_[idx];
    }
}
```

a-zakir replied:
you are right,

flomnes commented on src/io/inputs/model-converter/convertorVisitor.cpp:
```diff
@@ -389,32 +397,137 @@ std::any ConvertorVisitor::visitDual(ExprParser::DualContext* context)
+{
+    auto exprContexts = context->right_expr();
+    auto toStringTreeCallBack = [&context]() { return context->toStringTree(true); };
+    return processPower(exprContexts, toStringTreeCallBack);
```
```suggestion
    return processPower(exprContexts, context->toStringTree(true));
```

a-zakir replied:
The string won’t be used most of the time

flomnes replied:
I'm not sure I understand your argument

a-zakir replied:
You want to replace the callback with a string for each call. The string will contain the tree structure and will only be used to provide more detailed error messages.

a-zakir replied:
i'm avoiding the call to toStringTree for every time this operator is used.

flomnes commented on src/solver/optim-model-filler/ReadLinearExpressionVisitor.cpp:
```diff
@@ -288,18 +288,85 @@ Antares::Optimization::TimeDependentLinearExpression ReadLinearExpressionVisitor
+        return handlePow(node);
+    default:
+        // TODO
+        throw std::runtime_error("");
```
TODO

flomnes commented on src/solver/optim-model-filler/ReadLinearExpressionVisitor.cpp:
```diff
@@ -288,18 +288,85 @@ Antares::Optimization::TimeDependentLinearExpression ReadLinearExpressionVisitor
+    auto result(dispatch(operands.at(0)));
+    for (int i = 1; i < operands.size(); ++i)
+    {
+        const auto* operand = operands.at(i);
```
I think `std::vector` has an `operator[] const`, but I may be wrong

a-zakir replied:
yes if you don't want bound check [ ] is equivalent to .at()

flomnes commented on src/solver/optim-model-filler/ReadLinearExpressionVisitor.cpp:
```diff
@@ -288,18 +288,85 @@ Antares::Optimization::TimeDependentLinearExpression ReadLinearExpressionVisitor
+  const Nodes::FunctionNode* node)
+{
+    const auto& operands = node->getOperands();
+    // we know that max has at least two child
```
```suggestion
    // we know that max has at least two children
```

flomnes commented on src/expressions/include/antares/expressions/visitors/TimeIndexVisitor.h:
```diff
@@ -57,8 +58,10 @@ class TimeIndexVisitor: public NodeVisitor<Optimisation::TimeIndex>
     Optimisation::TimeIndex visit(const Nodes::AllTimeSumNode* timeSumNode) override;
-    Optimisation::TimeIndex visit(const Nodes::ReducedCostNode* node) override;
-    Optimisation::TimeIndex visit(const Nodes::DualNode* node) override;
+    Optimisation::TimeIndex handleReducedCost(const Nodes::FunctionNode* node);
+    Optimisation::TimeIndex handleDual(const Nodes::FunctionNode* node);
+    Optimisation::TimeIndex handlePow(const Nodes::FunctionNode* node);
```
These functions don't really need to be inside the class, they could be free functions

a-zakir replied:
free functions with the same names in different visitors break unity build

flomnes commented on src/expressions/include/antares/expressions/visitors/LinearStatus.h:
```diff
@@ -175,4 +175,15 @@ constexpr LinearStatus operator-(LinearStatus a)
+{
+    a = a & b;
+    return a;
+}
```
It seems that `operator&` really looks like `operator+`. Is there any reason to "duplicate" this operator ?

a-zakir replied:
the implementation is the same  
```
constexpr LinearStatus operator&(LinearStatus a, LinearStatus b)
{
    return operator+(a, b);
}

```
i don't want people to read linearStatus of max (a,b) = LinearStatus(a)+LinearStatus(b)

a-zakir replied:
code is updated

flomnes commented on src/expressions/include/antares/expressions/visitors/EvalVisitor.h:
```diff
@@ -322,7 +328,26 @@ class EvalVisitor: public NodeVisitor<EvaluationResult>
+            result = result.applyOperation(dispatch(operand), op);
+        }
+        return result;
+    }
```
I like that, it seems like another visitor has the same code, but with some duplication (min, max, etc.)

flomnes commented on src/expressions/include/antares/expressions/visitors/EvalVisitor.h:
```diff
@@ -20,11 +20,11 @@
 #include <sstream>
 #include <variant>
 
+#include <antares/expressions/nodes/FunctionNode.h>
```
Shouldn't be necessary, it comes from ExpressionsNodes.h

flomnes commented on src/expressions/include/antares/expressions/nodes/FunctionNode.h:
```diff
@@ -22,45 +22,59 @@
+    {
+    }
+
+    explicit FunctionNode(FunctionNodeType type, const std::vector<Node*>&& operands):
```
```suggestion
    explicit FunctionNode(FunctionNodeType type, std::vector<Node*>&& operands):
```

flomnes reviewed (commented):
Maybe document the procedure "how to add a new function node" ?

a-zakir committed:
pr reviews

flomnes committed:
Remove callback, use polymorhphic pointer instead (#3222)

flomnes committed:
Use checkMessage from test_utils_unit (#3223)

Instead of creating lambda functions, use the existing `checkMessage`
mechanism to check the string returned by `exception.what()`.

flomnes committed:
Fix a few compilation warnings (#3226)

a-zakir committed:
upd

a-zakir committed:
Merge remote-tracking branch 'origin/feature/functionNode-max' into feature/functionNode-max

# Conflicts:
#	src/solver/optim-model-filler/LinearExpression.cpp
#	src/solver/optim-model-filler/ReadLinearExpressionVisitor.cpp

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
add tests

a-zakir committed:
friend not needed

a-zakir committed:
add tests

a-zakir committed:
add tests

a-zakir committed:
format

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
add tests

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
upd

a-zakir committed:
fix smell code

a-zakir committed:
update

a-zakir committed:
update

a-zakir committed:
format

a-zakir committed:
Merge branch 'develop' into feature/functionNode-max

a-zakir committed:
upd

a-zakir committed:
upd

flomnes commented on src/expressions/include/antares/expressions/visitors/EvalVisitor.h:
```diff
@@ -122,7 +126,12 @@ class EvaluationResult
+    // [[nodiscard]] std::variant<double, std::vector<double>> value() const
+    // {
+    //     return value_;
+    // }
```
```suggestion
```

flomnes commented on src/expressions/include/antares/expressions/visitors/EvalVisitor.h:
```diff
@@ -280,6 +290,34 @@ EvaluationResult EvaluationResult::evaluateUnaryOperation(Op op) const
+
+    for (size_t i = 0; i < size; ++i)
+    {
+        for (size_t j = 0; j < in.size(); ++j)
```
I see two nested loops, it's not clear what `i` and `j` represent

flomnes reviewed (commented):
You only have time-independent tests, it would be nice to have at least a few tests with more than one timestep

a-zakir committed:
add test

a-zakir committed:
upd

a-zakir committed:
update

a-zakir committed:
update

a-zakir commented:
> You only have time-independent tests, it would be nice to have at least a few tests with more than one timestep
see
src/tests/src/expressions/test_PrintAndEvalVisitors.cpp
 src\tests\src\solver\optim-model-filler\test_componentFiller.cpp
